### DISCUSSION TITLE
Feature Request: Deliver Notifications Without Database Records

### DISCUSSION THREAD

**Discussion by viktor-shmigol** [excid3/noticed#375] [Ideas]
### **Problem**
I am currently in the process of upgrading my application to use Noticed v2. Within our notification system, there are specific notifications that do not require storage in the database. Due to the high volume of these notifications (thousands per day), we prefer not to track them in the database for performance reasons. Previously, we did not explicitly specify the deliver_by :database option. Consequently, no database records were created.
```
class DraftSentNotification < Noticed::Base
  deliver_by :email, mailer: 'UserMailer', method: :draft_sent, if: :email_notifications?
end
```
After upgrading to v2, I have not identified any option, either in the documentation or the code, that allows me to skip the creation of noticed_events and noticed_notifications records.

### Request
Is it possible to introduce a new configuration setting that allows notifications to be sent without generating corresponding database records? For example:
```
DraftSentNotification.deliver(user, deliver_database: false)
```
Such an enhancement would significantly benefit our use case by optimizing performance and reducing unnecessary database writes for specific types of notifications.

### Consideration
Without this capability, I would be forced to implement a daily job to remove these records from the database, which is not an ideal solution due to its complexity and potential impact on system resources.

Thank you for your consideration and assistance with this matter.

**Comment by excid3** [comment]
The database records fix a lot of issues such as sending bulk notifications (a notification to 1,000s of users now inserts using `insert_all`)

If you're on Postgres, have you seen unlogged tables? https://www.crunchydata.com/blog/postgresl-unlogged-tables
There may be equivalents for MySQL, etc.

Can you provide a better example than an email notification where you might need this? If it's just email, you should use ActionMailer directly instead of Noticed. So a more complex example would be handy to brainstorm ways this could be supported.

**Reply by viktor-shmigol** [reply]
To provide further clarity, our application utilizes this gem for dispatching notifications through various providers. In the previous version, no records were generated in the notifications table unless we explicitly used the deliver_by :database method.

In transitioning to the new version of the gem, we anticipate maintaining the same behavior. However, there appears to be no mechanism for delivering notifications while bypassing the creation of records in the noticed_events and noticed_notifications tables in the database. It is crucial for us to abstain from populating these records in the database, and conventional methods such as insert_all do not offer a viable solution.

Here's an illustrative example using the gem v1:

```
class NewMessageNotification < Noticed::Base
  deliver_by :internal_notification, class: 'DeliveryMethods::InternalNotification', if: :internal_notifications?
  deliver_by :sms
  deliver_by :email
  deliver_by :slack
end
```
Notably, the option deliver_by: :database is absent. Consequently, notifications are sent using different delivery methods without creating any records in the database.

Your attention to this matter is greatly appreciated!

**Reply by MrKirat** [reply]
@viktor-shmigol What have you done on your project with this issue?

**Reply by viktor-shmigol** [reply]
@MrKirat Hello, you may inherit your notifier from Noticed::Ephemeral .
E,g::
```
class NewMessageNotifier < Noticed::Ephemeral
  deliver_by :internal_notification, class: 'DeliveryMethods::InternalNotification'

  notification_methods do
    def channel
      "user_#{recipient.id}"
    end

    def payload
      { body: }
    end

    def body
      I18n.t('notifications.new_message', author: params[:message].author_name)
    end
  end
end
```
See: https://github.com/excid3/noticed/pull/382#issuecomment-1911798860

**Reply by MrKirat** [reply]
@viktor-shmigol Thank you!

**Comment by MrKirat** [comment]
Hi @excid3 !

I actually have the same issue. 

I'm using the noticed v1 on the project and considering upgrading to version 2. I found in the upgrade guide that starting from version 2 all the notifications will be saved into the database. At the moment, we have a lot of intense Notifications that have only the ActionCable delivery method. I assume, that upgrading to version 2 can significantly decrease the performance of the application because of increasing database interactions.

Can I disable saving into DB anyhow?

**Reply by excid3** [reply]
That's what the Ephemeral notifications are for. Inherit from `Noticed::Ephemeral` instead of `Noticed::Base`

**Reply by MrKirat** [reply]
@excid3  Thank you for the quick reply!
